### PR TITLE
Moving transform data out of Mat4

### PR DESCRIPTION
**Pull request by MarekLg**
Concerning #229.

As pointed out @AThilenius, having mutable transform data stored in a matrix results in rotation overlapping with scale and therefore affecting each other over time (see also [this article](https://www.gamasutra.com/view/news/173606/Indepth_Matrices_rotation_scale_and_drifting.php)).

This PR aims to resolve only this issue.
Any feedback and propositions for changes are very welcome!

### CODE REVIEW

File: crates/bevy_transform/src/transform_propagate_system.rs [hunk 1]
```diff
@@ -29,8 +29,8 @@ fn propagate_recursive(
             transform_query.get::<Transform>(entity),
             transform_query.get_mut::<GlobalTransform>(entity),
         ) {
-            *global_transform.value_mut() = parent * *transform.value();
-            *global_transform.value()
+            global_transform.value = *parent * transform.compute_matrix();
```
Comment (AThilenius) [inline crates/bevy_transform/src/transform_propagate_system.rs:32]: This requires recomputing the local space transform for every decedent of a parent recursively, which is unnecessary. The original impl cached it in `LocalToParent`. This also can never work with on-change detection because you are forced to recompute every hierarchy in case a descendant changed. Meaning every run you must recompute local space transforms, and global space transforms for every singe entity in a hierarchy regardless of if they changed or not. Especially in development builds where static objects in a hierarchy haven't yet been backed into a `GlobalTransform`, that sounds like an unacceptable performance overhead to me?

Comment (MarekLg) [reply crates/bevy_transform/src/transform_propagate_system.rs:32]: You're right, I totally overlooked that. Thanks for pointing that out!

Comment (MarekLg) [reply crates/bevy_transform/src/transform_propagate_system.rs]: Now the matrix is only recomputed when the transform changed. This still recomputes all the `GlobalTransform`s every frame, regardless if any changes occured, but this is how the system worked before and should IMO be for a different PR to fix.

File: crates/bevy_transform/src/components/transform.rs [hunk 2]
```diff
@@ -1,167 +1,184 @@
...
 impl Transform {
     #[inline(always)]
-    pub fn new(value: Mat4) -> Self {
-        Transform { value }
+    pub fn identity() -> Self {
+        Transform {
+            translation: Vec3::zero(),
+            rotation: Quat::identity(),
+            scale: Vec3::one(),
+            matrix_cache: Some(Mat4::identity()),
```
Comment (bitshifter) [inline crates/bevy_transform/src/components/transform.rs:20]: Caching the matrix in this transform is going to result in a cache miss for any access of translation, rotation or scale. Rust automatically orders struct fields by alignment and Mat4 is 16 byte aligned due to SIMD so it will go first. It's a 64 byte struct so it will fill an entire cache line by itself. Additionally putting it inside an Option will add an addition 15 bytes of padding to the beginning of the structure. An `Option<Mat4>` will start with a discriminant byte that tells Rust if the option is `Some` or `None`, this byte will be aligned to the same alignment of what the option holds, so 16, this the 15 bytes of padding.

It would be more efficient to keep these cached values in a separate structure. However if you want to keep it in the same structure you would be better off getting rid of the `Option` and just use a `bool` to flag if there is a cache value or not.

Also `glam` has a type which stores transforms as translation, rotation (as `Quat`) and scale so you could use that instead of writing it yourself. It does require enabling the `transform-types` feature of `glam` though.

Comment (cart) [reply crates/bevy_transform/src/components/transform.rs:20]: Yeah I don't think we should store the "matrix cache" here because the type is now "too big"

But I also want to point out that if we rip out the matrix into a separate component, we now have re-created the very "synchronization problem" that we were trying to solve in the first place with the transform rework. At that point, why _not_ use individual components like @AThilenius's initial implementation used? Then we get the benefits of "better cache perf and easier access to individual parts of a transform". 

I think its very hard to justify a Transform { pos/rot/scale } component w/ a matrix cache (either in another component or embedded inside Transform). Unless I'm missing something, we should either eat the "mat4 creation cost" at the call site whenever a full 4x4 matrix is required, or we should just move back to the old system (or something like it).

Comment (cart) [reply crates/bevy_transform/src/components/transform.rs:20]: I think I like @termhn's idea to use Similarities and multiply those directly whenever possible. We only really need a 4x4 when writing data used by shaders, which only needs to happen once per rendered GlobalTransform.

https://github.com/bevyengine/bevy/issues/229#issuecomment-699172675

Comment (termhn) [reply crates/bevy_transform/src/components/transform.rs:20]: Note that `Transform { pos/rot/scale }` (with no matrix cache) *is* a Similarity (in glam known as a [`TransformSRT`](https://github.com/bitshifter/glam-rs/blob/master/src/f32/transform.rs#L12)). But I agree, the transformation of that to a matrix should only happen as @cart described

Comment (MarekLg) [reply crates/bevy_transform/src/components/transform.rs:20]: Those are some very good points.

> Unless I'm missing something, we should either eat the "mat4 creation cost" at the call site whenever a full 4x4 matrix is required, or we should just move back to the old system (or something like it).

I think that is the main question we have to decide on. I'm personally against separate components, only for the fact that we can take away more work from the user (the reason being explained in the lower part of [my comment](https://github.com/bevyengine/bevy/issues/229#issuecomment-699610211)).

Comment (AThilenius) [reply crates/bevy_transform/src/components/transform.rs:20]: A few holes I would like to poke (seems like my MO at this point lol). Firstly let me drop [this link](https://community.amethyst.rs/t/scaling-transforms-and-hierarchies-oh-my/1159) here to gently and in the most passive-aggressive way possible suggest that `legion_transform` was pretty well iterated on and thought out (although never completed mostly because of problems with Legion at the time). Not perfect certainly, but history is currently (in the issue thread and this PR) repeating itself.

- This implementation will never play nice with change detection. When you update pos/rot/scale it's picked up as a change. When you then re-compute the local-space matrix, it will be picked up as another change. As it stands now that will cause another full hierarchy recompute. Even if you use a dirty flag, you still need to enumerate over unchanged entities and conditionally branch (in a way that Branch Prediction units are bad at too). Ergo... it should be in its own component. But...
- You don't need a `LocalSpace` component because that's redundant for root entities. Local space is the exact same thing a global space. You only need local space when the entity is a child, and that local space is local to it's parent. Perhaps you could call that component `LocalToParent`? 😉
- You 'could' avoid computing a local-space mat4 for children, with some of the Projective GA @termhn was talking about. But multiplying Similarities directly is slower than multiplying matricies. So that's a net loss on perf. I ran this test a while ago, so [this post](https://community.amethyst.rs/t/mechanism-to-sub-dispatch-group-of-systems/1197/8?u=xercsess) where I compared Isometries (easier to multiply than Similarities) and matricies. Matricies win by a healthy margin and that was with nalgebra which doesn't have any SIMD support IIRC. I would very much support some exploration down the Projective GA route though, but multiplying Similarities is a bad idea.
- I don't actually mind pos/rot/scale being combined 🤷‍♂️ I don't think it's useful, especially when Hecs has Bundles, and it will cause performance loss, but it won't make or break anything. However... scale should not be stored as non-uniform by default. That is a bad idea.

Comment (termhn) [reply crates/bevy_transform/src/components/transform.rs:20]: > You 'could' avoid computing a local-space mat4 for children, with some of the Projective GA @termhn was talking about. But multiplying Similarities directly is slower than multiplying matrices. So that's a net loss on perf. I ran this test a while ago, see this post where I compared Isometries (easier to multiply than Similarities) and matrices. Matrices win by a healthy margin and that was with nalgebra which doesn't have any SIMD support IIRC. I would very much support some exploration down the Projective GA route though, but multiplying Similarities is probably not the way to go.

Respectfully, I think this microbenchmark isn't really representative... I highly suspect that for whatever reason, the nalgebra similarity code wasn't getting properly inlined or optimized. When I do the benchmark for ultraviolet Mat4*Mat4 vs Similarity3*Similarity3 in mathbench-rs (using ultraviolet rather than glam just 'cause the mathbench stuff is already implemented for them), I get Mat4*Mat4 at 13ns and Sim3*Sim3 at 15ns on my machine... so the Similarities are indeed slightly slower, but not by much

Comment (termhn) [reply crates/bevy_transform/src/components/transform.rs:20]: > However... scale should not be stored as non-uniform by default. That is a bad idea.

Agreed

Comment (cart) [reply crates/bevy_transform/src/components/transform.rs:20]: > However... scale should not be stored as non-uniform by default. That is a bad idea.
I totally agree that we should encourage uniform scales. But to me, the way we do that isn't as clear cut. We could create apis that encourage users to use uniform scales, but still store them as non-uniform. That way switching between uniform/non-uniform doesn't require archetype changes (or multiple component Bundle structs with uniform/nonuniform variants like Sprite/SpriteNonUniformScale)

From a performance perspective, I have no clue and would want benchmarks to inform that decision.

Comment (AThilenius) [reply crates/bevy_transform/src/components/transform.rs:20]: @termhn is making me question my performance fortitude, so I'll thumbs up implementing this PR as an `Isometry/Similarity/Affine` and skipping the `LocalToParent` mat4 caches until there is a more realistic benchmark 🤷‍♂️

Out of curiosity I did collect this from the game we are making in Unity DOTs a few days ago (I can't say much about the game other than it's well into it's production phase).
Transform Update Analysis

The TL;DR is that a lot of entities don't move at all which isn't surprising. Of the ones that do, only half-ish have local-space changes (which are almost entirely skeletal animations). But I need to run this on a build when I have free time not just in-editor. That data would at least guide a more realistic performance bench.

Comment (bitshifter) [reply crates/bevy_transform/src/components/transform.rs:20]: > However... scale should not be stored as non-uniform by default. That is a bad idea.

I am going to go against popular opinion here. On balance I don't see any value in special casing uniform scale. In my experience artists, level designers, tech designers etc. use non uniform scale. From the content creation point of view uniform scale doesn't seem any more or less common or useful than other forms of non-uniform scaling. A simple use case of a non uniform scale would be reflecting an object over one or two axes. It seems to me that if the long term goal of bevy is an easy to use engine for non programmers to use then it is going to have to support non-uniform scale. Telling content people not to use it is not going to get you very far and for good reason - there are lots of valid uses for it on the content side, e.g. instancing a lot of geometry into a level. Physics is a bit of a special case, but the number of objects that are simulated is usually a small proportion of objects in a scene. Also physics engines like PhysX will support non-uniform scale on meshes (see https://documentation.help/NVIDIA-PhysX-SDK-Guide/Shapes.html#mesh-scaling), again because instancing is important. It's only really some primitives that you can't scale non-uniformly.

tl;dr non-uniform scale is an important feature that should be supported, special casing uniform scale adds more code complexity and you still need to deal with the non-uniform case. So what benefit is there in separating them?

Component wise I can see benefits in making scale optional (e.g. bundle position and rotation together but scale separate), but separate uniform and non-uniform not so much.

Comment (cart) [reply crates/bevy_transform/src/components/transform.rs:20]: I (currently) agree with @bitshifter. Non-uniform scale is useful in a variety of contexts. The physics case is one to be wary of (and potentially error check), but I'm not yet convinced that we need to provide hard protections against it.

My preference right now is to do one of the following (or something like them):

## No Distinction, public fields
```rust
struct Transform {
  pub translation: Vec3,
  pub rotation: Quat,
  pub scale: Vec3,
}
```
* pros: dead simple, interacts nicely with the bevy "properties" api 
* cons: no inherent protection against non-uniform scale precision errors or physics breakage 

## Uniform scale is "scale", non-uniform scale is "non_uniform_scale", private fields
```rust
struct Transform {
  translation: Vec3,
  rotation: Quat,
  scale: Scale,
}

enum Scale {
  Uniform(f32),
  NonUniform(Vec3),
}

impl Transform {
  pub fn set_scale(&mut self, scale: f32) {}
  pub fn scale(&self) -> Option<f32> {} /* might not be uniform */ 
  pub fn non_uniform_scale(&self) -> Vec3 {}
  pub fn set_non_uniform_scale(&mut self, scale: Vec3) {}
}
```
* pros: encourages "uninformed" people looking for scale to "do the right thing' by default
* cons: more cumbersome / more api overhead. harder to interact with because you cant do things like `*transform.scale *= 3.0`. does not interact nicely with bevy "properties" api (although this kind of case needs a workaround at some point anyway)

Comment (termhn) [reply crates/bevy_transform/src/components/transform.rs:20]: I think I'm convinced. I think the public option is definitely the more attractive one -- it we're going to support nonuniform as first class then might as well make it truly first class.

Comment (cart) [reply crates/bevy_transform/src/components/transform.rs:20]: Cool. I think I would prefer a custom Transform type with the relevant methods/traits implemented instead of new-typing a math-lib type. It makes directly property access nicer and this logic isnt particularly complicated.

Lets wait a day or so for someone to chime in with counterpoints. Otherwise I'm inclined to move forward on

```rust
struct Transform {
  pub translation: Vec3,
  pub rotation: Quat,
  pub scale: Vec3,
}
```

Comment (MarekLg) [reply crates/bevy_transform/src/components/transform.rs:20]: I'll revert the last commit so everyone checking in is on the same page.

Comment (cart) [reply crates/bevy_transform/src/components/transform.rs]: Cool. I think the biggest missing piece at this point is implementing Transform x Transform = Transform and using that when calculating GlobalTransform. To get the most out of this, I think we should also make GlobalTransform a similarity (ideally not a newtype to allow convenient direct field access). But that then creates the issue of "how do we convert GlobalTransform to matrix bytes when creating uniform buffers?". I think "data-driven uniforms" need a better story for "this type should be converted to this other byte-convertible type when used as a uniform". The current system assumes the type is directly byte-convertible.

In the short term I think something like this will work:

```rust
#[derive(RenderResources, RenderResource)]
#[render_resources(from_self)]
pub struct GlobalTransform {
    pub translation: Vec3,
    pub rotation: Quat,
    pub scale: Vec3,
}

impl Bytes for GlobalTransform {
    fn write_bytes(&self, buffer: &mut [u8]) {
        // convert self to matrix and write bytes to buffer
    }

    fn byte_len(&self) -> usize {
        return std::mem::size_of::<[f32; 16]>();
    }
}
```

I think it also makes sense to remove things like the "rotate" and "apply_scale" helper methods in favor of direct field access.

Comment (MarekLg) [reply crates/bevy_transform/src/components/transform.rs]: I fully agree with the first part.

```rust
    pub translation: Vec3,
    pub rotation: Quat,
    pub scale: Vec3,
```

I'm not quite sold on the attributes being public though, yet. The way I'm perceiving `GlobalTransform`, the developer shouldn't / shouldn't need to mutate it themselves, or am I forgetting something?

> I think it also makes sense to remove things like the "rotate" and "apply_scale" helper methods in favor of direct field access.

I think this comes down to the question "how much should the developer know to use the engine". Most people should find it trivial to apply one `Quat` to another, but I think the few added lines of code are a worthwhile price for not requiring that sort of knowledge. It could be, however, that I'm reading too much into the priorities of this project, in which case I'd be totally fine with that change.

File: crates/bevy_transform/src/components/transform.rs [hunk 3]
```diff
@@ -1,167 +1,132 @@
...
+        Transform {
+            rotation,
+            ..Default::default()
+        }
     }
 
     pub fn from_scale(scale: f32) -> Self {
-        Transform::new(Mat4::from_scale(Vec3::splat(scale)))
+        Transform {
+            scale: Vec3::one() * scale,
```
Comment (bitshifter) [inline crates/bevy_transform/src/components/transform.rs:48]: You could do Vec3::splat(scale) here so it's just assignment with no multiplication.

File: crates/bevy_transform/src/components/transform.rs [hunk 4]
```diff
@@ -1,167 +1,132 @@
...
+        self.scale = scale;
         self
     }
 
-    pub fn with_apply_scale(mut self, scale: f32) -> Self {
-        self.apply_scale(scale);
-        self
+    /// Returns transform with the same translation and scale, but rotation so that transform.forward() points at the origin
+    pub fn looking_at_origin(self) -> Self {
+        self.looking_at(Vec3::zero(), Vec3::unit_y())
```
Comment (bitshifter) [inline crates/bevy_transform/src/components/transform.rs:90]: As an aside it might pay for bevy to have some constants for `UP`, `FORWARD` and `SIDE` (or `RIGHT`) axes, might make it clearer what the coordinate system is.

Comment (bitshifter) [reply crates/bevy_transform/src/components/transform.rs:90]: I've added support for creating constants in `glam` but I need to publish a new release. It's a bit tricky because underlying intrinsics aren't `const fn` so macros have to be used.

File: crates/bevy_transform/src/components/transform.rs (hunk 2 above)
Comment (cart) [reply crates/bevy_transform/src/components/transform.rs]: > I'm not quite sold on the attributes being public though, yet. The way I'm perceiving GlobalTransform, the developer shouldn't / shouldn't need to mutate it themselves, or am I forgetting something?

I see no harm in making them public in this case. First, it provides parity with Transform (which I think is the biggest win). People can access "translation" the same way on both components. Different apis would increase the cognitive load. Additionally, there may be cases where people want to directly modify GlobalTransform (ex: given that the engine internally uses GlobalTransform as the "actual transform of the entity", there might eventually be a case where we have an entity that only has GlobalTransform).

> I think this comes down to the question "how much should the developer know to use the engine". Most people should find it trivial to apply one Quat to another, but I think the few added lines of code are a worthwhile price for not requiring that sort of knowledge. It could be, however, that I'm reading too much into the priorities of this project, in which case I'd be totally fine with that change.

I think I'd like to keep it simple for now. Every api we add has a cost. If it turns out that this is a real usability gap, we can always layer it on top.

File: crates/bevy_transform/src/components/transform.rs (hunk 4 above)
Comment (cart) [reply crates/bevy_transform/src/components/transform.rs:90]: Ooh thats cool / much appreciated :smile:

**Comment by cart** [comment]
I'd like to include this in the next Bevy release (which should be in the next week or so). This is pretty close to being ready. @MarekLg do you want me to push the last few requested changes needed or would you prefer to do it? Don't feel pressured to do work you don't have time for. You've already done a ton of good work here :heart:

**Comment by MarekLg** [comment]
Feel free to make the final changes. I don't have much time this weekend, sadly. But thanks for the kind words, looking forward to more contributions!

**Comment by cart** [comment]
Alrighty I made GlobalTransform a similarity, implemented math ops (using @termhn's Similarity impl as a reference), and removed some methods from Transform/GlobalTransform to match my personal taste.

I'll leave this open for a day or two in case someone wants to comment on my changes, then its merge time!
